src/openforms/forms/tasks.py

```diff
@@ -87,6 +105,33 @@ def recouple_submission_variables_to_form_variables(form_id: int) -> None:
         logger.info("Form variables were updated while this task was runnning.")
 
 
+@app.task(ignore_result=True)
+def recouple_reusable_definition_variables_to_form_variables(form_id: int) -> None:
+    """
+    When the FormVariable bulk create/update endpoint is called, all existing FormVariable related to the form are
+    deleted and new are created. If there are any form definitions which are reusable, we want to update all the forms
+    which are also using these FormDefinitions (concerning the FormVariables). This task updates the FormVariables
+    of each related form in the database, by deleteing the old ones and recreating them (the most safe way).
+    """
+    from .models import FormDefinition, FormVariable  # due to circular import
+
+    form = Form.objects.get(id=form_id)
+    reusable_form_definitions = FormDefinition.objects.filter(
+        formstep__form=form, is_reusable=True
+    ).distinct()
+
+    for fd in reusable_form_definitions:
+        form_ids_to_update = fd.formstep_set.exclude(form=form).values_list(
+            "form", flat=True
+        )
+
+        for form_id in form_ids_to_update:
+            form_to_update = Form.objects.get(id=form_id)
+            form_to_update.formvariable_set.all().delete()
+
+            FormVariable.objects.create_for_form(form_to_update)
```

Comment:
I think you can simplify and optimize this:

```py
fds = FormDefinition.objects.filter(formstep__form=form_id, is_reusable=True)  # 0 queries
other_forms = Form.objects.filter(formstep__form_definition__in=fds).exclude(id=form_id)  # 0 queries
# delete the existing form variables, we will re-create them
FormVariable.objects.filter(form__in=other_forms, source=FormVariableSources.component).delete()  # 1 query
for form in other_forms:  # 2 queries
    FormVariable.objects.create_for_form(form)  # x number of queries for each form
```

We can avoid of nested loops (but you'll have to mark the task as `@transaction.atomic`!

The logic is:

1. Figure out which re-usable form definitions are involved
2. Get a collection of all the other forms that use any of these re-usable form definitions
3. Delete all their component form variables
4. Regenerate the component form variables for each form